design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,292 @@
+`CertificateSigningRequests` are cluster scoped resources, the signer name must
+include the namespace, if the referenced `Issuer` is namespaced. The signer name
+for cert-manager signers will be prefixed with `cert-manager.io` to prevent
+conflicts with other external signer projects.
+
+```yaml
+  # Namespaced issuer reference
+  signerName: cert-manager.io/<namespace>.<issuer-name>
+
+  # Cluster scoped issuer reference
+  signerName: cert-manager.io/<issuer-name>
+```
+
+Using the same approach of referencing by _just_ name, rather than issuer type
+(e.g. CA, Vault etc.), keeps the behaviour of this resource in line with
+`CertificateRequests` for end users.
+
+Each `CertificateSigningRequest` controller will behave in the same way as the
+existing `CertificateRequest` resource, by getting the referenced signer, and
+attempting to sign. If the issuer type is not managed by this controller, do
+nothing, else sign.
+
+
+### Signers
+
+Some special cases for some `[Cluster]Issuers` that need to be addressed:
+
+- SelfSigned: Makes use of annotations, and so these annotations should also be
+    present on `CertificateSigningRequests`.
+
+- Venafi: Makes use of annotations, and so these annotations should also be
+    present on `CertificateSigningRequests`.
+
+- ACME: Makes use of annotations, and so these annotations should also be
+    present on `CertificateSigningRequests`.
+
+- ACME: The ACME controller creates sub-resources (`Orders`). Since
+    `CertificateSigningRequests` are cluster scoped resources, when referencing
+    a `ClusterIssuer` we should create `Orders` in the `Cluster Resource
+    Namespace` (default `cert-manager`), else the namespace of the referenced
+    `Issuer`.
+
+
+### Conditions
+
+The `CertificateSigningRequest` has [well-known condition
+types](https://github.com/kubernetes/api/blob/4a626d306b987a4096cf0784ec01af1be2f6d67f/certificates/v1/types.go#L222)
+of `Approved`, `Denied`, and `Failed`. All `CertificateSigningRequest` signer
+controllers should not begin computation of the request until the resource has
+the `Approved` condition set to `True`.
+
+In the absence of further policy, the `Approved` condition is always set with
+`True`, _unless_, the `CertificateSigningRequest` references a namespaced
+`Issuer` and the requester does not have create permissions for
+`certificates.k8s.io/CertificateSigningRequests` in that namespace. In the case
+of this namespace validation failing, the `CertificateSigningRequest` `Denied`
+condition should be set to `True`, and no signer controllers should process the
+request.
+
+If a signer fails when processing a `CertificateSigningRequest`, the well-known
+`Failed` condition should be set to `True`. No signer controllers should process
+the request again.
+
+If the referenced issuer resource doesn't exist or is not in a ready state, the
+corresponding `CertificateSigningRequest` controller should set the `Pending`
+condition to `True`.
```

Comment:
`Pending` condition type would be specific to cert-manager, is that correct?

Reply:
We have decided not to go with the Pending condition.